packages/core-app-api/src/apis/implementations/auth/pinniped/Pinniped.ts

```diff
@@ -0,0 +1,198 @@
+    const {
+      configApi,
+      environment = 'development',
+      provider = DEFAULT_PROVIDER,
+      oauthRequestApi,
+      discoveryApi,
+      defaultScopes = [
+        'openid',
+        'offline_access',
+        'pinniped:request-audience',
+        'username',
+      ],
+      popupOptions,
+    } = options;
+
+    this.configApi = configApi;
+    this.environment = environment;
+    this.provider = provider;
+    this.oauthRequestApi = oauthRequestApi;
+    this.discoveryApi = discoveryApi;
+    this.defaultScopes = defaultScopes;
+    this.popupOptions = popupOptions;
+
+    this.audiences = {};
+  }
+
+  async getSupervisorIdToken(
+    _options?: AuthRequestOptions | undefined,
+  ): Promise<string> {
+    throw new Error('Method not implemented.');
+  }
+
+  async getClusterScopedIdToken(
+    audience: string,
+    options?: AuthRequestOptions | undefined,
+  ): Promise<string> {
+    const aud = await Pinniped.getAudience(audience);
+    if (!(aud in this.audiences)) {
+      const connector = new DefaultAuthConnector({
+        configApi: this.configApi,
+        discoveryApi: this.discoveryApi,
+        environment: this.environment,
+        provider: this.provider,
+        oauthRequestApi: this.oauthRequestApi,
+        sessionTransform({
+          backstageIdentity,
+          ...res
+        }: OAuth2Response): OAuth2Session {
+          const session: OAuth2Session = {
+            ...res,
+            providerInfo: {
+              idToken: res.providerInfo.idToken,
+              accessToken: res.providerInfo.accessToken,
+              scopes: Pinniped.normalizeScopes(
+                scopes => scopes,
+                res.providerInfo.scope,
+              ),
+              expiresAt: res.providerInfo.expiresInSeconds
+                ? new Date(
+                    Date.now() + res.providerInfo.expiresInSeconds * 1000,
+                  )
+                : undefined,
+            },
+          };
+          return session;
+        },
+        popupOptions: this.popupOptions,
+      });
+
+      const sessionManager = new RefreshingAuthSessionManager({
+        connector,
+        defaultScopes: new Set(this.defaultScopes),
+        sessionScopes: (session: OAuth2Session) => session.providerInfo.scopes,
+        sessionShouldRefresh: (session: OAuth2Session) => {
+          // TODO(Rugvip): Optimize to use separate checks for provider vs backstage session expiration
+          let min = Infinity;
+          if (session.providerInfo?.expiresAt) {
+            min = Math.min(
+              min,
+              (session.providerInfo.expiresAt.getTime() - Date.now()) / 1000,
+            );
+          }
+          if (session.backstageIdentity?.expiresAt) {
+            min = Math.min(
+              min,
+              (session.backstageIdentity.expiresAt.getTime() - Date.now()) /
+                1000,
+            );
+          }
+          return min < 60 * 5;
+        },
```

Comment:
I think we know this isn't needed since neither the [authenticate](https://github.com/backstage/backstage/blob/30e76f392e3cb84071438c13088d5bec8d4ea92f/plugins/auth-backend-module-pinniped-provider/src/authenticator.ts#L210-L219) nor the [refresh](https://github.com/backstage/backstage/blob/30e76f392e3cb84071438c13088d5bec8d4ea92f/plugins/auth-backend-module-pinniped-provider/src/authenticator.ts#L248-L257) method on the authenticator actually returns a backstage identity.
```suggestion
```